includes/Blocks/Block.php

```diff
@@ -366,73 +366,60 @@ private function register_type(): void {
 	 */
-	private function resolve_block_attributes_recursive( $attributes, $html, $config ): array {
+	private function resolve_block_attributes_recursive( $attributes, string $html, array $config ): array {
 		$result = [];
+
 		foreach ( $config as $key => $value ) {
 			// Get default value.
 			$default = $value['default'] ?? null;
 			$source  = $value['source'] ?? null;
+
 			switch ( $source ) {
 				case 'rich-text':
 				case 'html':
+					// If there is no selector, we are dealing with single source.
 					if ( ! isset( $value['selector'] ) ) {
 						$result[ $key ] = $this->parse_single_source( $html, $source );
-					} else {
-						$result[ $key ] = DOMHelpers::parseHTML( $html, $value['selector'] );
-	
-						if ( isset( $value['multiline'] ) && ! empty( $result[ $key ] ) ) {
-							$result[ $key ] = DOMHelpers::getElementsFromHTML( $result[ $key ], $value['multiline'] );
-						}
+						break;
 					}
+
+					$result[ $key ] = $this->parse_html_source( $html, $value );
 					break;
 				case 'attribute':
-					$result[ $key ] = DOMHelpers::parseAttribute( $html, $value['selector'], $value['attribute'], $default );
+					$result[ $key ] = $this->parse_attribute_source( $html, $value );
 					break;
 				case 'text':
-					$result[ $key ] = DOMHelpers::parseText( $html, $value['selector'] );
+					$result[ $key ] = $this->parse_text_source( $html, $value );
 					break;
 				case 'query':
-					$temp  = [];
-					$nodes = DOMHelpers::findNodes( $html, $value['selector'] );
-
-					// Coerce nodes to an array if it's not already.
-					if ( ! is_array( $nodes ) ) {
-						$nodes = [ $nodes ];
-					}
-	
-					foreach ( $nodes as $source_node ) {
-						foreach ( $value['query'] as $q_key => $q_value ) {
-							$temp_config    = [
-								$q_key => $q_value,
-							];
-							$res            = $this->resolve_block_attributes_recursive( $attributes, $source_node->html(), $temp_config );
-							$temp[ $q_key ] = $res[ $q_key ];
-						}
-						$result[ $key ][] = $temp;
-					}
+					$result[ $key ] = $this->parse_query_source( $html, $value, $attributes );
 					break;
 			}
+
 			// Post processing of return value based on configured type
-			switch ( $value['type'] ) {
-				case 'integer':
-					$result[ $key ] = intval( $result[ $key ] );
-					break;
-				case 'boolean':
-					if ( false === $result[ $key ] ) {
+			if ( array_key_exists( $key, $result ) ) {
+				switch ( $value['type'] ) {
+					case 'integer':
+						$result[ $key ] = intval( $result[ $key ] );
 						break;
-					}
-					if ( is_null( $result[ $key ] ) ) {
-						$result[ $key ] = false;
+					case 'boolean':
+						if ( false === $result[ $key ] ) {
+							break;
+						}
+						if ( is_null( $result[ $key ] ) ) {
+							$result[ $key ] = false;
+							break;
+						}
+						$result[ $key ] = true;
 						break;
-					}
-					$result[ $key ] = true;
-					break;
+				}
 			}
-			
+
+			// Fallback to the attributes or default value if the result is empty.
```

Comment:
Now that this method is cleaned up, we can more easily see that the `$result[ $key ]` is getting replaced with the fallback or default value even when it probably shouldn't: 
E.g. an `int` value of 0, or a boolean value of `false`.

I didn't want to change this behavior because while I doubt it's intentional, I'm not 100% sure and didn't want to introduce a breaking change.

Reply:
@theodesp @ChrisWiegman when one of you has a minute can you confirm whether this is a bug or intended behavior?